modules/asset-manager-workflowoperation/src/main/java/org/opencastproject/workflow/handler/assetmanager/SelectVersionWorkflowOperationHandler.java

```diff
@@ -0,0 +1,201 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+package org.opencastproject.workflow.handler.assetmanager;
+
+import org.opencastproject.assetmanager.api.AssetManager;
+import org.opencastproject.assetmanager.api.Snapshot;
+import org.opencastproject.assetmanager.api.query.AQueryBuilder;
+import org.opencastproject.assetmanager.api.query.ARecord;
+import org.opencastproject.assetmanager.api.query.AResult;
+import org.opencastproject.job.api.JobContext;
+import org.opencastproject.mediapackage.MediaPackage;
+import org.opencastproject.mediapackage.MediaPackageElement;
+import org.opencastproject.mediapackage.MediaPackageElementFlavor;
+import org.opencastproject.mediapackage.selector.SimpleElementSelector;
+import org.opencastproject.workflow.api.AbstractWorkflowOperationHandler;
+import org.opencastproject.workflow.api.WorkflowInstance;
+import org.opencastproject.workflow.api.WorkflowOperationException;
+import org.opencastproject.workflow.api.WorkflowOperationInstance;
+import org.opencastproject.workflow.api.WorkflowOperationResult;
+import org.opencastproject.workflow.api.WorkflowOperationResult.Action;
+
+import com.entwinemedia.fn.data.Opt;
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+/**
+ * Replaces the media package in the current workflow with a previous version from the asset manager. There are two ways
+ * to choose the version: by version number or a combination of source-flavors and no-tags (choose the latest version
+ * where elements with source-flavor do not have the tags specified in no-tags.
+ *
+ * This operation should be the first one in a workflow executed from the archive because it REPLACES the media package
+ * used by the current workflow.
+ */
+public class SelectVersionWorkflowOperationHandler extends AbstractWorkflowOperationHandler {
+
+  private static final Logger logger = LoggerFactory.getLogger(SelectVersionWorkflowOperationHandler.class);
+
+  public static final String OPT_VERSION = "version";
+  public static final String OPT_NO_TAGS = "no-tags";
+  public static final String OPT_SOURCE_FLAVORS = "source-flavors";
+
+  /** The configuration options for this handler */
+  private static final SortedMap<String, String> CONFIG_OPTIONS;
+
+  private AssetManager assetManager;
+
+  static {
+    CONFIG_OPTIONS = new TreeMap<String, String>();
+    CONFIG_OPTIONS.put(OPT_VERSION,
+            "Select the specific version. If informed, takes precedence over the other options.");
```

Comment:
From the docs, I would've expected the WOH to fail if both of mutually exclusive way of selecting a version are present. To avoid unexpected behaviour that might be hard to debug, I would either change this to actually fail if mutually exclusive ways of selecting a version are specified, or instead mention the precedence order in the docs.

Reply:
Good point!